dandiapi/api/models/audit.py

```diff
@@ -0,0 +1,198 @@
+        *, dandiset: Dandiset, user: User, removed_owners: list[User], added_owners: list[User]
+    ):
+        def glean_user_info(user: User):
+            return {
+                'username': user.username,
+                'email': user.email,
+                'name': f'{user.first_name} {user.last_name}',
+            }
+
+        details = {
+            'removed_owners': [glean_user_info(u) for u in removed_owners],
+            'added_owners': [glean_user_info(u) for u in added_owners],
+        }
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='change_owners', details=details
+        )
+
+    @staticmethod
+    def update_metadata(*, dandiset: Dandiset, user: User, metadata: dict) -> AuditRecord:
+        details = {'metadata': metadata}
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='update_metadata', details=details
+        )
+
+    @staticmethod
+    def _asset_details(asset: Asset) -> dict:
+        checksum = (
+            (asset.blob and asset.blob.etag)
+            or (asset.embargoed_blob and asset.embargoed_blob.etag)
+            or (asset.zarr and asset.zarr.checksum)
+        )
+
+        return {
+            'path': asset.path,
+            'asset_blob_id': asset.blob and asset.blob.id,
+            'embargoed_asset_blob_id': asset.embargoed_blob and asset.embargoed_blob.id,
+            'zarr_archive_id': asset.zarr and asset.zarr.id,
+            'asset_id': asset.id,
+            'checksum': checksum,
+            'metadata': asset.metadata,
+        }
+
+    @staticmethod
+    def add_asset(*, dandiset: Dandiset, user: User, asset: Asset) -> AuditRecord:
+        details = AuditRecord._asset_details(asset)
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='add_asset', details=details
+        )
+
+    @staticmethod
+    def update_asset(*, dandiset: Dandiset, user: User, asset: Asset) -> AuditRecord:
+        details = AuditRecord._asset_details(asset)
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='update_asset', details=details
+        )
+
+    @staticmethod
+    def remove_asset(*, dandiset: Dandiset, user: User, asset: Asset) -> AuditRecord:
+        details = {
+            'path': asset.path,
+            'asset_id': asset.id,
+        }
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='remove_asset', details=details
+        )
+
+    @staticmethod
+    def create_zarr(*, dandiset: Dandiset, user: User, zarr_archive: ZarrArchive) -> AuditRecord:
+        details = {
+            'zarr_id': zarr_archive.id,
+            'name': zarr_archive.name,
+        }
+        return AuditRecord.make_audit_record(
+            dandiset=dandiset, user=user, record_type='create_zarr', details=details
+        )
+
+    @staticmethod
+    def upload_zarr(
```

Comment:
If we're going to name the deletion operation `delete_zarr_chunks`, I think it makes sense to name this operation `upload_zarr_chunks`. That would make it both consistent as well describe the operation more accurately.

Reply:
:point_up_2:

Reply:
Done in 7f21061e75ed57af72420c832e8844bc31161853.